goship.go

```diff
@@ -760,23 +778,210 @@ func extractOutputHandler(fn func(http.ResponseWriter, *http.Request, string, st
+
+			authUrl, err := provider.GetBeginAuthURL(state, nil)
+			if err != nil {
+				http.Error(w, err.Error(), http.StatusInternalServerError)
+				return
+			}
+
+			// redirect
+			http.Redirect(w, r, authUrl, http.StatusFound)
+		}
+	} else {
+		return func(w http.ResponseWriter, r *http.Request) {}
+	}
+}
+
+func callbackHandler(providerName string, auth bool) http.HandlerFunc {
+	// Handle error
+	if auth == true {
+		provider, err := gomniauth.Provider(providerName)
+		if err != nil {
+			panic(err)
+		}
+		return func(w http.ResponseWriter, r *http.Request) {
+
+			omap, err := objx.FromURLQuery(r.URL.RawQuery)
+			if err != nil {
+				log.Printf("error getting resp from callback")
+				http.Error(w, err.Error(), http.StatusInternalServerError)
+				return
+			}
+
+			creds, err := provider.CompleteAuth(omap)
+			if err != nil {
+				http.Error(w, err.Error(), http.StatusInternalServerError)
+				return
+			}
+
+			user, userErr := provider.GetUser(creds)
+			if userErr != nil {
+				log.Printf("Failed to get user from Github %s", user)
+				http.Error(w, err.Error(), http.StatusInternalServerError)
+				return
+			}
+
+			session, err := store.Get(r, session_name)
+			if err != nil {
+				log.Fatal("Can't get a session store")
+			}
+
+			session.Options = &sessions.Options{
+				Path:     "/",
+				MaxAge:   86400 * 7,
+				HttpOnly: true,
+			}
+
+			session.Values["userName"] = user.Nickname()
+			session.Values["avatarURL"] = user.AvatarURL()
+			log.Print("saving session")
+			session.Save(r, w)
+
+			http.Redirect(w, r, os.Getenv("GITHUB_CALLBACK_URL"), http.StatusFound)
+
+		}
+	} else {
+		return func(w http.ResponseWriter, r *http.Request) {}
+	}
+}
+
+type auth struct {
+	authorization       bool
+	githubRandomHashKey string
+	githubOmniauthID    string
+	githubOmniauthKey   string
+	githubCallbackURL   string
+}
+
+func getAuth() auth {
+	a := auth{}
+	a.githubRandomHashKey = os.Getenv("GITHUB_RANDOM_HASH_KEY")
+	a.githubOmniauthID = os.Getenv("GITHUB_OMNI_AUTH_ID")
+	a.githubOmniauthKey = os.Getenv("GITHUB_OMNI_AUTH_KEY")
+	a.githubCallbackURL = os.Getenv("GITHUB_CALLBACK_URL")
+	a.authorization = true
+	// Let user know if a key is missing and that auth is disabled.
+	if a.githubRandomHashKey == "" || a.githubOmniauthID == "" || a.githubOmniauthKey == "" || a.githubCallbackURL == "" {
+		log.Printf("Missing one or more Gomniauth Environment Variables: Running with with limited functionality! \n githubRandomHashKey [%s] \n githubOmniauthID [%s] \n githubOmniauthKey[%s] \n githubCallbackURL[%s]",
+			a.githubRandomHashKey,
+			a.githubOmniauthID,
+			a.githubOmniauthKey,
+			a.githubCallbackURL)
+		a.authorization = false
+	} else {
+		gomniauth.SetSecurityKey(a.githubRandomHashKey)
+		gomniauth.WithProviders(
+			githubOauth.New(a.githubOmniauthID, a.githubOmniauthKey, a.githubCallbackURL+"/auth/github/callback"),
+		)
+		a.authorization = true
+	}
+	return a
+}
+
+func isCollaborator(owner, repo, user string) bool {
```

Comment:
I'd like to see some comments that explain what these functions do, even if they are private :lollipop: :baby: